Add Smoke and Foam chemical reaction effects.

Radrark opened the pull request:
Draft because I have not added the foam effect yet, but the smoke one is fully functional:
- It spreads in a range proportional to the reaction size.
- It consumes all the reagents from the solution it was created from.
- It repeatedly applies the touch effect of every reagent it has to every entity it's covering.
- It gradually copies the reagents it has to every bloodstreams of entities it's covering, except those with internals on.

The foam effect will be similar. Any comments or requests are appreciated.

collinlunn commented on Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs:
```diff
@@ -0,0 +1,61 @@
+
+        public void React(IEntity solutionEntity, double intensity)
+        {
+            var contents = solutionEntity.GetComponent<SolutionContainerComponent>();
```
Instead of using GetComponent, which throws an exception if that component is not there, use TryGetComponent, report an error with Logger.Error then return out, so that it doesn't crash the server if something is messed up. Same for GetComponent< SmokeComponent >().

collinlunn commented on Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs:
```diff
@@ -0,0 +1,61 @@
+            if (!mapManager.TryFindGridAt(solutionEntity.Transform.MapPosition, out var grid)) return;
+
+            var coords = grid.MapToGrid(solutionEntity.Transform.MapPosition);
+            var ent = solutionEntity.EntityManager.SpawnEntity("smoke_cloud", coords.SnapToGrid());
```
Preferably use a prototype field instead of a specific prototype, so multiple smoke could prototypes can be used.

collinlunn commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+            if (_running)
+                return;
+            TryAddSolution(solution);
+            Start(Owner, amount, duration+2*amount*spreadDelay, spreadDelay, duration+amount*spreadDelay);
```
Its not immediately clear why `lifeTime` and `averageLifeTime` are calculated with those formulas, some explaining comments would help.

collinlunn commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+                    if (neighbor.TryGetComponent(out AirtightComponent airtight) && airtight.AirBlocked)
+                        return;
+                }
+                var newSmoke = Owner.EntityManager.SpawnEntity("smoke_cloud", snapGrid.DirectionToGrid(dir));
```
Again would like to not be hardcoded to a specific prototype

collinlunn commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+                if (!entity.TryGetComponent(out BloodstreamComponent bloodstream))
+                    continue;
+
+                if (entity.TryGetComponent(out InternalsComponent internals) &&
+                    internals.BreathToolEntity != null &&
+                    internals.GasTankEntity != null &&
+                    internals.BreathToolEntity.TryGetComponent(out BreathToolComponent breathTool) &&
+                    breathTool.IsFunctional &&
+                    internals.GasTankEntity.TryGetComponent(out GasTankComponent gasTank) &&
+                    gasTank.Air != null)
+                    continue;
```
Add a property to InternalsComponent  that answers "Will this person breath in chems" instead of doing the checks in the SmokeComponent

collinlunn reviewed (commented):
There are spaces missing after commas in various method calls, please clean those up.

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+            var solutionFraction = 1 / Math.Floor(_averageLifetime / ExposeDelay);
+
+            // Reagents react with the tile under the smoke
+            foreach (var reagentQuantity in contents.ReagentList.ToArray())
+            {
+                if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                reagent.ReactionTile(tile, reagentQuantity.Quantity * solutionFraction);
+            }
+
+            foreach (var entity in tile.GetEntitiesInTileFast())
+            {
+                // Reagents touch all entities on the tile
+                foreach (var reagentQuantity in contents.ReagentList.ToArray())
+                {
+                    if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                    var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                    reagent.ReactionEntity(entity, ReactionMethod.Touch, reagentQuantity.Quantity * solutionFraction);
+                }
```
Going through ReagentList thrice seems redundant, you could probably combine all 3 (at the very least these 2).

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+                    gasTank.Air != null)
+                    continue;
+
+                var cloneSolution = contents.Solution.Clone();
+                var transferAmount = ReagentUnit.Min(cloneSolution.TotalVolume * solutionFraction, bloodstream.EmptyVolume);
+                var transferSolution = cloneSolution.SplitSolution(transferAmount);
```
Same thing on cloning.

Radrark replied:
In SS13 the reagents in the smoke don't get consumed, otherwise smoke would be useless against crowded areas.

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+                    transferSolution.RemoveReagent(reagentQuantity.ReagentId,reagent.ReactionEntity(entity, ReactionMethod.Ingestion, reagentQuantity.Quantity));
+                }
+
+                bloodstream.TryTransferSolution(transferSolution);
```
With no cloning you'd need to add it back into the smoke if the transfer failed.

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+            if (!result)
+                return;
+
+            if (Owner.TryGetComponent(out SpriteComponent sprite))
+                sprite.Color = contents.SubstanceColor;
```
Is this expensive? Using sprite states pushes all of the data over the network and doing this for a lot of entities could be a lot of bandwidth.

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+            }
+        }
+
+        private void Start(IEntity inception, int amount, float lifetime, float spreadDelay, float averageLifetime)
+        {
+            _inception = inception;
+            _amount = amount;
+            _lifetime = lifetime;
+            _spreadDelay = spreadDelay;
+            _averageLifetime = averageLifetime;
+            _running = true;
+
+            Owner.SpawnTimer((int) (_spreadDelay*1000), SpreadSmoke);
+        }
```
Just make them properties and have whatever spawns it deal with setting them IMO.

Also roll the timer into the update as mentioned elsewhere.

moonheart08 replied:
IMO this should be kept as it makes refactoring in the future easier

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+            if (!_running)
+                return;
+
+            if (IsInception)
+            {
+                if (_exposeTimer > ExposeDelay)
+                {
+                    _exposeTimer = 0;
+                    ReactWithTileAndEntities();
+                }
+                _exposeTimer += frameTime;
+            }
+            else
+            {
+                if (_inception.TryGetComponent(out SmokeComponent smokeComp) &&
+                    smokeComp._exposeTimer > ExposeDelay)
+                {
+                    ReactWithTileAndEntities();
+                }
+            }
```
Rather than chucking all of this into the component IMO you're better off having the system keep track of which components are connected which should also have better perf.

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+
+        public void Update(float frameTime)
+        {
+            if (!_running)
+                return;
```
I doubt this would ever not be true?

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+
+        public override string Name => "Smoke";
+
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
```
If you moved the reaction stuff onto the system you could avoid having to store these references on every smoke component.

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+            {
+                if (_exposeTimer > ExposeDelay)
+                {
+                    _exposeTimer = 0;
```
When this is moved make it -= frameTime to avoid being tickrate dependent.

metalgearsloth reviewed (changes requested):
tl;dr move some more stuff onto the system mainly.

Radrark commented on Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:
```diff
@@ -0,0 +1,210 @@
+                var newSmoke = Owner.EntityManager.SpawnEntity("smoke_cloud", snapGrid.DirectionToGrid(dir));
+                var smokeComponent = newSmoke.GetComponent<SmokeComponent>();
+
+                if (Owner.TryGetComponent(out SolutionContainerComponent contents))
+                {
+                    var solution = contents.Solution.Clone();
+                    smokeComponent.TryAddSolution(solution);
+                }
```
In SS13 the reagents get copied to every smoke cloud. The foam effect is the one that dilutes the reagents.

Radrark commented on Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs:
```diff
@@ -0,0 +1,61 @@
+            var coords = grid.MapToGrid(solutionEntity.Transform.MapPosition);
+            var ent = solutionEntity.EntityManager.SpawnEntity("smoke_cloud", coords.SnapToGrid());
+
+            var smokeComponent = ent.GetComponent<SmokeComponent>();
+            smokeComponent.Activate(solution, amount, _duration, _spreadDelay);
```
But how would I add the solution and set the amount using Initialize? They need get added by the thing that spawns the smoke.

Radrark commented:
Refactored a bit and did all you said except: 
- Changing the cloning of solutions. It works like that in SS13 and I think is fine.
-  Setting netsync of the sprite component to false. I think it's pointless to make a SmokeVisualizer for only changing the color of the sprite.

metalgearsloth commented:
> 
> 
> Refactored a bit and did all you said except:
> 
>     * Changing the cloning of solutions. It works like that in SS13 and I think it's fine.
> 
>     * Setting netsync of the sprite component to false. I think it's pointless to make a SmokeVisualizer for only changing the color of the sprite.

Yeah the first feels icky but it is how tg does it so it's okay for now
Setting the sprite once is fine for now but the main issue with it is it will send the entire state over when anything changes.

Rest seems good.

metalgearsloth commented:
Did you want to merge this now or add in foam first?

Radrark commented:
I'm gonna add the foam first.

Radrark commented:
Ok so as I was adding the foam I noticed I was copypasting a lot of code and that can't be good right? So I refactored my code like 3 times and got this. Also I added a smoke visualizer. It should be good now, ready for review.

metalgearsloth reviewed (changes requested):
I read through the tg foam code to have a better understanding of what's going on as some stuff is unintuitive. Despite the number of comments they're mostly really minor changes and I'm pretty happy with it ingame IMO.

Also should give audio attribution somewhere (in a comment is fine).

Radrark commented on Resources/Prototypes/Entities/Effects/chemistry_effects.yml:
```diff
@@ -0,0 +1,191 @@
+  id: FoamedIronMetal
+  name: foamed iron metal
+  parent: BaseFoamedMetal
+  components:
+    - type: Sprite
+      sprite: Effects/foam.rsi
+      state: ironfoam
+    - type: Icon
+      sprite: Effects/foam.rsi
+      state: ironfoam
```
1. Yes for the entity spawn panel if someone wants to spawn them.
2. The parent entity has it false.

metalgearsloth replied:
Entity spawn panel doesn't show abstract+ icon can be pulled from the sprite now. We're trying to phase out iconcomponent where possible and the panel will also use visualizer defaults.

Radrark commented on Content.Server/GameObjects/Components/Chemistry/FoamComponent.cs:
```diff
@@ -0,0 +1,158 @@
+
+        private void HandleKillMessage()
+        {
+            if (Owner.TryGetComponent(out AppearanceComponent appearance))
+            {
+                appearance.SetData(FoamVisuals.State, true);
+            }
+            Owner.SpawnTimer(600, () =>
```
To solve this it needs prediction right? I'm not sure if I could do that.

Radrark commented on Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:
```diff
@@ -0,0 +1,248 @@
+                    if (neighbor.TryGetComponent(out AirtightComponent airtight) && airtight.AirBlocked)
+                        return;
+                }
+                var newEffect = Owner.EntityManager.SpawnEntity(Owner.Prototype?.ID, snapGrid.DirectionToGrid(dir));
+                if (!newEffect.TryGetComponent(out AreaEffectComponent effectComponent))
+                {
+                    newEffect.Delete();
+                    return;
+                }
+
+                effectComponent.Amount = Amount - 1;
+                Inception.Add(effectComponent);
+
+                spawned.Add(newEffect);
+            }
+
+            SpreadToDir(Direction.North);
+            SpreadToDir(Direction.East);
+            SpreadToDir(Direction.South);
+            SpreadToDir(Direction.West);
```
Does that change anything? These 4 calls are made in the same frame right?

metalgearsloth replied:
Ahh I'm dumb I thought the amount was being adjusted each time. You would think the tiles it spreads to would be the same if it's in a corridor or open space so it is weird.

Radrark commented on Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:
```diff
@@ -0,0 +1,248 @@
+
+    }
+
+    public class AreaEffectInception
+    {
+        private const float ReactionDelay = 0.5f;
+
+        private readonly HashSet<AreaEffectComponent> _group = new();
+
+        private float _lifeTimer;
+        private float _spreadTimer;
+        private float _reactionTimer;
```
And make an entity prototype that has this inception component right?

Radrark commented on Content.Server/Chemistry/ReactionEffects/FoamReactionEffect.cs:
```diff
@@ -0,0 +1,118 @@
+            if (!mapManager.TryFindGridAt(solutionEntity.Transform.MapPosition, out var grid)) return;
+
+            var coords = grid.MapToGrid(solutionEntity.Transform.MapPosition);
+            var ent = solutionEntity.EntityManager.SpawnEntity(_foamPrototypeId, coords.SnapToGrid());
```
I think it's okay to have two different foams overlap.

metalgearsloth replied:
Okey dokey.

Radrark commented:
Addressed all the comments except the ones I replied.

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:
```diff
@@ -0,0 +1,259 @@
+
+    }
+
+    [RegisterComponent]
+    public class AreaEffectInceptionComponent : Component
```
Chuck in its own file now.

metalgearsloth commented on Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:
```diff
@@ -0,0 +1,259 @@
+    {
+        public override string Name => "AreaEffectInception";
+
+        private const float ReactionDelay = 0.5f;
+
+        private readonly HashSet<AreaEffectComponent> _group = new();
+
+        private float _lifeTimer;
+        private float _spreadTimer;
+        private float _reactionTimer;
+
+        private int _amountCounterSpreading;
+        private int _amountCounterRemoving;
+
+        private float _duration;
+        private float _spreadDelay;
+        private float _removeDelay;
+        private float _averageExposures;
```
Could also add VV to these if you wanted now.

PaulRitter replied:
some xmldocs pls

metalgearsloth reviewed (commented):
Rest LGTM, good work! Just a comment on the spawn panel.

PaulRitter commented on Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs:
```diff
@@ -0,0 +1,71 @@
+            }
+
+            _foamDissolve = new Animation {Length = TimeSpan.FromSeconds(delay)};
+            {
+                var flick = new AnimationTrackSpriteFlick();
+                _foamDissolve.AnimationTracks.Add(flick);
+                flick.LayerKey = FoamVisualLayers.Base;
+                flick.KeyFrames.Add(new AnimationTrackSpriteFlick.KeyFrame(state, 0f));
+            }
```
why create a new scope here?

PaulRitter commented on Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs:
```diff
@@ -0,0 +1,71 @@
+            {
+                if (state)
+                {
+                    var animPlayer = component.Owner.GetComponent<AnimationPlayerComponent>();
```
use trygetcomponent here

PaulRitter commented on Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs:
```diff
@@ -0,0 +1,71 @@
+
+            if (component.TryGetData<Color>(FoamVisuals.Color, out var color))
+            {
+                var sprite = component.Owner.GetComponent<ISpriteComponent>();
```
use trygetcomponent here

PaulRitter commented on Content.Client/GameObjects/Components/Chemistry/SmokeVisualizer.cs:
```diff
@@ -0,0 +1,21 @@
+
+            if (component.TryGetData<Color>(SmokeVisuals.Color, out var color))
+            {
+                var sprite = component.Owner.GetComponent<ISpriteComponent>();
```
use trygetcomponent here

PaulRitter commented on Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs:
```diff
@@ -0,0 +1,71 @@
+
+namespace Content.Client.GameObjects.Components.Chemistry
+{
+    public class FoamVisualizer : AppearanceVisualizer
```
add a [UsedImplicitly]

PaulRitter commented on Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs:
```diff
@@ -0,0 +1,132 @@
+            if (!_mapManager.TryFindGridAt(solutionEntity.Transform.MapPosition, out var grid)) return;
+
+            var coords = grid.MapToGrid(solutionEntity.Transform.MapPosition);
+            var ent = solutionEntity.EntityManager.SpawnEntity(_prototypeId, coords.SnapToGrid());
```
default value for _prototypeId is null, so this will just spawn an entity without any components if no prototype is provided :/ either check in ExposeData if a prototype was provided or handle it accordingly here

PaulRitter commented on Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs:
```diff
@@ -0,0 +1,132 @@
+    {
+        [Dependency] private readonly IMapManager _mapManager = default!;
+
+        /// <summary>
+        /// Used for calculating the spread range of the smoke based on the intensity of the reaction.
+        /// </summary>
+        private float _rangeConstant;
+        private float _rangeMultiplier;
+
+        private int _maxRange;
+
+        private bool _diluteReagents;
+        private int _reagentDilutionStart;
+        private float _reagentDilutionFactor;
+        private float _reagentMaxConcentrationFactor;
+        /// <summary>
+        /// How many seconds will the smoke stay, counting after fully spreading.
+        /// </summary>
+        private float _duration;
+        /// <summary>
+        /// How many seconds between each spread step.
+        /// </summary>
+        private float _spreadDelay;
+
+        private float _removeDelay;
+
+        private string _prototypeId;
+        private string _sound;
```
more xmldocs would probably be appropriate here

PaulRitter commented on Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:
```diff
@@ -0,0 +1,259 @@
+                return;
+
+            // Make every outer square from the group spread
+            if (_amountCounterSpreading != 0)
```
nitpick: use `>`

PaulRitter commented on Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:
```diff
@@ -0,0 +1,259 @@
+                _reactionTimer -= ReactionDelay;
+                foreach (var effect in _group)
+                {
+                    effect.React(_averageExposures);
```
check that effect isn't already deleted, since effect.Kill() in line 202 will trigger the such

Radrark replied:
effect.Kill() already removes the effect from the group so I don't think this is necessary

PaulRitter replied:
correct

PaulRitter reviewed (changes requested):
smokecomponent & foamcomponent NEED to be generalized & not copy pasted. if you want support on this, @ me in discord (username: /paul)

AJCM-git commented on Resources/Prototypes/Recipes/Reactions/chemicals.yml:
```diff
@@ -57,6 +71,94 @@
+      removeDelay: 0.5
+      diluteReagents: false
+      prototypeId: Smoke
+      sound: /Audio/Effects/smoke.ogg #Taken from https://github.com/tgstation/tgstation/blob/master/sound/effects/smoke.ogg
```
Dont put licenses inside the prototype, there should be an license.txt inside the sound folder, put it in there

ike709 commented:
Conflicts

PaulRitter reviewed (changes requested):
Short foreword before we get started:
I know this must be painful since technically you already have working code, but sadly working is not equal futureproof. (Trust me i've had to learn that too)
In order for future coders to easily implement new AreaEffects or fix bugs that will eventually pop up, its is crucial for us to write code which uses proper abstraction & inheritance. If we were to copypaste two classes, one would fix a bug in the one class & would have to remember to do it in the other one too. Having our logic in one place only is very important to keep [technical debt](https://en.wikipedia.org/wiki/Technical_debt) down. Anyways, enough said, please forgive me for torturing you with this, here we go:


More concrete Requirements of what i'd like to see:
- Rename `AreaEffectComponent` to `SolutionAreaEffectComponent`, makes more sense imo.
- Make `AreaReactionEffect` abstract.
- defer implementation of lines 143-148 of `AreaReactionEffect` (getting of the areaeffectcomponent) to a abstract method `protected abstract SolutionAreaEffectComponent GetAreaEffectComponent(IEntity entity)`
- Create two inheritors of `AreaReactionEffect`: `FoamAreaReactionEffect` & `SmokeAreaReactionEffect`. Implement `GetAreaEffectComponent` in here, leave it empty for now, we will implement it later, need to rework something first.

- Make `SolutionAreaEffectComponent` abstract. This will make sense later.
- Merge `AreaEffectInceptionComponent` and `SolutionAreaEffectComponent` into one Component. The split doesn't make sense imo and just makes it a bit more confusing
- Merge `FoamComponent` and `SmokeComponent` into `SolutionAreaEffectComponent`, and here is how:
  - Logic from `HandleSpreadMessage` and `TryAddSolution` can just be merge
  - Do NOT merge in the logic of `HandleKillMessage`! Rather define a new abstract method `protected abstract void OnKill()` and call it whenever you'd send the KillMessage
  - Now for the big whoppa: `HandleReactMessage`:
    - Afaict the only difference between the two methods here is the last part, the entity interaction.
    - Therefore, to merge it into `SolutionAreaEffectComponent`, merge the code up to last foreach-loop. now define a new abstract method `protected abstract void ReactWithEntity(IEntity entity)`. call this method in the last foreach loop instead of doing all you logic
- Now create two inheritors of `SolutionAreaEffectComponent`: `FoamSolutionAreaEffectComponent` and `SmokeSolutionAreaEffectComponent`
- implement the abstract methods with the logic we left out earlier when merging it into `SolutionAreaEffectComponent`

- Now we loop back to `FoamAreaReactionEffect` and `SmokeAreaReactionEffect`: Implement `GetAreaEffectComponent` by trying to get the corresponding component (either `FoamSolutionAreaEffectComponent` or `SmokeSolutionAreaEffectComponent`)

Done! Now test it out and @ me on discord cause you eventually found out i made a major error when bikeshedding this problem.

Radrark commented:
Ok first of all thanks for the mini tutorial it must've taken you some time and it helped me a lot. So, I did all the things except merging AreaEffectInceptionComponent and SolutionAreaEffectComponent because I'm receiving conflicting orders, metalgearsloth told me in an earlier review to make the inception component. So by the silicon policy, I'm able to choose what order to follow and I'll keep the inception component :P, I think it makes things simpler and it uses the already implemented infrastructure.

PaulRitter commented on Content.Server/GameObjects/Components/Chemistry/SolutionAreaEffectComponent.cs:
```diff
@@ -0,0 +1,188 @@
+        [ComponentDependency] protected readonly SnapGridComponent? SnapGridComponent = default!;
+        [ComponentDependency] protected readonly SolutionContainerComponent? SolutionContainerComponent = default!;
+        public int Amount { get; set; }
+        public SolutionAreaEffectInceptionComponent? Inception { get; set; }
```
the inceptioncomponent does not need to be an actual component. also the ticking should probably just be moved into this class.

What you should do is:
- Move `InceptionUpdate` into this class
- Replace the component with a `SolutionAreaEffectGroup`-Class that still has the add/remove functions. The newly ported over function then just takes the hashset from inside the solutionareaeffectgroup

PaulRitter replied:
DONT. FORGET WHAT I SAID

PaulRitter commented on Content.Server/Chemistry/ReactionEffects/SmokeAreaReactionEffect.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        protected override SolutionAreaEffectComponent GetAreaEffectComponent(IEntity entity)
+        {
+            return !entity.TryGetComponent(out SmokeSolutionAreaEffectComponent comp) ? null : comp;
```
cleaner to use `GetComponentOrNull`

PaulRitter commented on Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs:
```diff
@@ -78,7 +78,7 @@ public class AreaReactionEffect : IReactionEffect
         private string _sound;
 
-        public AreaReactionEffect()
+        protected AreaReactionEffect()
```
why do this?

Radrark replied:
idk the IDE suggested it. Makes sense though because AreaReactionEffect is an abstract class so there's no point in having a public constructor.

PaulRitter reviewed (changes requested):
looking very good! almost there

PaulRitter reviewed (changes requested):
Looks good so far.

Just gotta change it to use nullability, then you're golden for a merge :)

Radrark commented:
Done.

PaulRitter commented on Content.Server/GameObjects/EntitySystems/SolutionAreaEffectSystem.cs:
```diff
@@ -0,0 +1,18 @@
+﻿using Content.Server.GameObjects.Components.Chemistry;
```
```suggestion
#nullable enable
﻿using Content.Server.GameObjects.Components.Chemistry;
```
